sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -275,13 +275,16 @@ object PullupCorrelatedPredicates extends Rule[LogicalPlan] with PredicateHelper
     plan transformExpressions {
       case ScalarSubquery(sub, children, exprId) if children.nonEmpty =>
         val (newPlan, newCond) = pullOutCorrelatedPredicates(sub, outerPlans)
-        ScalarSubquery(newPlan, newCond, exprId)
+        val conds = newCond ++ children.filter(_.isInstanceOf[Predicate])
```

Comment:
I am not sure about this. What if we use children if `newCond` is empty?

Reply:
Hmmm... That might not be correct. Since in the analysis phase, `children` contains all outer references. Even though `newCond` is empty, we can't leave out `children` as it is.

Reply:
But, if there are outer references there, `newCond` should not be empty, right? I am a bit worried about using `Predicate` here. You might have an outer reference in a `Coalesce` or `CaseWhen`, which are not `Predicate`s for instance...

Reply:
That makes sense... Basically we want to distinguish analyzed plans and optimized plans. But in the current implementation, both the analyzer and optimizer are stripping outer references...

Reply:
Also, a side note... The `PullupCorrelatedPredicates` rule is tightly coupled with `RewriteSubqueryPredicates`. For `ListQuery`, it seems that `PullupCorrelatedPredicates` is compulsory for a correct physical plan.

Reply:
@yeshengm A couple of questions. 
1. Given this rule is run once, do we have a problem in real-life ? Or we just want to make sure    things would be okay if this got fired more than one time ?
2. I am also little hesitant on using Predicate check like @mgaido91. Should we check if the subplan has outer references and if not, just return the original expression ?

Reply:
@dilipbiswal We want to ensure all these optimizer rules can be run multiple times. This is critical for adaptive query execution.

Reply:
Also, for (2), it won't work for `ListQuery`. As you can see in the source code, the clause for `ListQuery` is executed even there's no outer reference. This is important for correctness...

Reply:
@gatorsmile 
>We want to ensure all these optimizer rules can be run multiple times. This is critical for adaptive query execution.

Got it. Thanks.. 

@yeshengm 
> Also, for (2), it won't work for ListQuery. As you can see in the source code, the clause for ListQuery is executed even there's no outer reference. This is important for correctness...

Yeah.. you are right. Queries having IN clauses may or may not have outer references.

Reply:
@dilipbiswal Could elaborate a bit on this issue? From my understanding, this rule `PullupCorrelatedPredicates` should only run if the plan contains outer ref. How can it generate wrong plans if a `ListQuery` does not contain outer refs?

Reply:
@yeshengm Trying to remember... 

> From my understanding, this rule PullupCorrelatedPredicates should only run if the plan contains outer ref.

I am not a 100% sure.. but from what i can see and remember, the reason we call PullupCorrelatedPredicates without a guard is probably to handle du-duplication of projected attributes. However, NOW the de-duplication is handled in other places as well. So we probably need to take another look at it and clean-up.

Reply:
@gatorsmile 

> We want to ensure all these optimizer rules can be run multiple times. This is critical for adaptive query execution.

Quick clarification question: does Spark 2.4.x perform double-optimization when adaptive execution is used? Or is that only a characteristic of the new / expanded adaptive execution features in Spark 3.x? I ask in order to help determine whether this patch should be a correctness backport for 2.4.x.

Reply:
@joshrosen-stripe Our adaptive execution in 2.4 or prior does not have this issue. It does not re-optimize the plans.

Reply:
FYI, I just checked that `dedupOuterReferencesInSubqueries` is in the analyzer rule and thus we can safely add the guard here. Also, we might consider combine `PullupCorrelatedPredicates` and `RewritePredicateSubqueries` since they are basically doing different parts of a same thing (and most importantly, applying the same hack on `SubqueryExpression`'s `children` field).

Reply:
https://github.com/apache/spark/blob/f8f4ec8d4d31968d2ae5fe6bdb51f0eea25a2c92/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala#L189 What I just proposed is a two-year-old TODO anyway...

Reply:
@yeshengm Yeah.. Actually i have been working on SPARK-19712. In the last pass of this work i was planning on merging these two rules into one. Thats why at the end of phase1 work .. i had marked it as a todo :-)

Reply:
@dilipbiswal Is there any particular reason making it hard to combine the two?

Reply:
@yeshengm So today, rewritePredicateSubquery happens almost at the end of optimization process where as PullupCorrelatedPredicates happens at the beginning of optimization process. In SPARK-19712, we are attempting to move rewritePredicateSubquery to the beginning of optimizer. Once that happens, i don't anticipate any problem in combining the two.

Reply:
@dilipbiswal According to jira, it seems that spark-19712 is already resolved. That said, we can safely combine the two now, right?

Reply:
@yeshengm 
> According to jira, it seems that spark-19712 is already resolved

Yeah :-). As you can see, there are multiple PRs against the JIRA. It got marked resolved after the first PR, i think. But we are doing this work in stages. I am working on couple of more prs to complete this work.

> That said, we can safely combine the two now, right?

To the best of my knowledge, we can't combine the two until we move rewrite to the beginning of optimization process (this is the piece i am working on at the moment).

Reply:
Ahh, got it. Thanks for your detailed explanation!

Reply:
@dilipbiswal Then I'll close this PR for now and wait till you combine those two rules. (or you can do that yourself). Thanks!

Reply:
@yeshengm Ok.. sounds good.

One question, can we not make this `PullupCorrelatedPredicates` idempotent now the way it is (i.e when these two rules are separate) ? If we did something like this : 
```
 case l @ ListQuery(sub, _, exprId, childOutputs) =>
        val (newPlan, newCond) = pullOutCorrelatedPredicates(sub, outerPlans)
        if (newCond.isEmpty) {
           // Perhaps just returning `l` may work as well. But in case we r relying on
           // the de-dup processing somehow..
           ListQuery(newPlan, l.children, exprId, childOutputs)
        } else {
           ListQuery(newPlan, newCond, exprId, childOutputs)
        }
```
will it work ? Or you tried it already ?

Reply:
Yes, I agree with your suggestion @dilipbiswal, basically it is what I suggested [earlier](https://github.com/apache/spark/pull/25164#discussion_r303609476)

Reply:
@mgaido91 Thanks a lot. Lets see if it works or if there is something we may be missing :-)

Reply:
I doubt it... Because the logic for checking `OuterReferences` and the logic for actually pulling up predicates are slightly different. With that being said, even though `l.children` is non-empty, it does not necessarily mean that `newCond` is empty.

The most natural way I can think of is that we combine these two rule `PullupCorrelatedPredicates` and `RewriteSubqueryPredicates`, `RewriteSubqueryPredicates` completely removes those hacky list subqueries. I don't think the plan can change if we apply these two rules in a single `Once` batch.

Reply:
TBH subquery resolution and optimzation are super tricky and can be error-prone. The current code is a bit complex and fragile, because one piece of code might have some pre-conditions on some other part of the codebase, which might change over time.

Reply:
@yeshengm Thanks .. let me make a pr to get it tested to see what is broken. I think it makes sense to keep this logic (even though its merged with rewriteSubquery) idempotent. That is because, pullupCorrelatedPredicates works for scalar subquery as well and that is handled by a different rule.